lib/presentation/widgets/list_characters.dart

```diff
@@ -1,56 +1,113 @@
 import 'package:flutter/material.dart';
+import 'package:flutter_bloc/flutter_bloc.dart';
+import 'package:marvelapp_flutter/presentation/features/home/bloc/home_bloc.dart';
+import 'package:marvelapp_flutter/presentation/features/home/bloc/home_event.dart';
 import 'package:marvelapp_flutter/presentation/models/character_view_data.dart';
 import 'package:marvelapp_flutter/presentation/navigation/app_routes.dart';
+import 'package:marvelapp_flutter/presentation/widgets/bottom_error.dart';
+import 'package:marvelapp_flutter/presentation/widgets/bottom_loader.dart';
+import 'package:marvelapp_flutter/presentation/widgets/end_of_list.dart';
 
-class ListCharacters extends StatelessWidget {
+class ListCharacters extends StatefulWidget {
   final List<CharacterViewData>? list;
+  final String? error;
+  final bool hasReachedMax;
 
-  const ListCharacters({Key? key, required this.list}) : super(key: key);
+  const ListCharacters({Key? key, required this.list, this.error, this.hasReachedMax = false}) : super(key: key);
+
+  @override
+  State<ListCharacters> createState() => _ListCharactersState();
+}
+
+class _ListCharactersState extends State<ListCharacters> {
+  final _scrollController = ScrollController();
+
+  @override
+  void initState() {
+    super.initState();
+    _scrollController.addListener(_onScroll);
+  }
 
   @override
   Widget build(BuildContext context) {
+    List<CharacterViewData> list = widget.list ?? List.empty();
     return Padding(
       padding: const EdgeInsets.all(6.0),
       child: ListView.builder(
-        itemCount: list?.length,
-        itemBuilder: (context, index) {
-          final item = list?[index] as CharacterViewData;
+        itemCount: list.length,
+        controller: _scrollController,
+        itemBuilder: (BuildContext context, int index) {
+          final item = list[index];
           String name = item.name ?? "";
-          return Card(
-            shape: const RoundedRectangleBorder(
-              borderRadius: BorderRadius.all(
-                Radius.circular(5.0),
-              ),
-            ),
-            child: InkWell(
-              onTap: () {
-                Navigator.pushNamed(context, AppRoutes.detailScreen, arguments: item.id);
-              },
-              child: Row(
-                children: [
-                  ClipRRect(
-                    borderRadius: const BorderRadius.only(
-                      topLeft: Radius.circular(5.0),
-                      bottomLeft: Radius.circular(5.0),
+          Widget errorOrLoader;
+          if (widget.error == null && !widget.hasReachedMax) {
+            errorOrLoader = const BottomLoader();
+          } else if (widget.error == null && widget.hasReachedMax) {
+            errorOrLoader = const EndOfList();
+          } else {
+            errorOrLoader = const BottomError();
+          }
+          return index >= list.length - 1
```

Comment:
`index > list.length - 1` should not be possible right?
And would not be replaced by the `EndOfList` widget the last element in the list?

Reply:
> index > list.length - 1 should not be possible right? 

 you mean "==" insted of >=    ?
in case index > list.length - 1  i can not see loading indicator
>And would not be replaced by the EndOfList widget the last element in the list?
 
i test this, yes. Last element replaced.
change all. Last element appeared in list. EndOfList change to EmptyWidget, but this widget we cant see and we dont need it.